classes/class-wp-tfpdf.php

```diff
@@ -0,0 +1,108 @@
+<?php
+/**
+ * To avoid modifying tFPDF directly we introduce WP_tFPDF,
+ * A thin wrapper around the parent class to utilize WP_Filesystem
+ */
+// phpcs:disable WordPress.WP.AlternativeFunctions.file_system_read_fwrite, WordPress.WP.AlternativeFunctions.file_system_read_fread, WordPress.WP.AlternativeFunctions.file_system_read_fopen, WordPress.WP.AlternativeFunctions.file_system_read_fclose, WordPress.VIP.FileSystemWritesDisallow
+if ( ! defined( 'ABSPATH' ) ) exit; // Exit if accessed directly
+
+require_once ABSPATH . '/wp-admin/includes/file.php';
+
+class WP_tFPDF extends tFPDF {
+	function __construct( $orientation = 'P', $unit = 'mm', $size = 'A4' ) {
+		parent::__construct( $orientation, $unit, $size );
+		$this->init_wp_filesystem();
+	}
+
+	function init_wp_filesystem() {
```

Comment:
I haven't used `WP_Filesystem` before, so I'm digging in and trying to understand how this will work for our users.

From my reading and testing, it seems that this initialization method will work fine in the following two cases:

1. We are using the `direct` method to access the filesystem.
2. We have the FTP creds stored in constants (e.g. in `wp-config.php`).

However, this **does not** work in the case where we need to prompt for credentials. In this case, we would need the prompt to happen in the context of a `POST` request so we can do the redirection properly. Unfortunately, at this point we do not have this `POST` request since there is no actual code path within the plugin that performs the PDF creation.

I wonder if the following would be a better solution:

- Do not use `WP_Filesystem` by default (i.e. go back to using the regular `tFPDF` class).
- Have a filter (and document it!) that enables the use of this subclass instead (perhaps `sensei_certificiates_use_wp_filesystem`).
- Inside this class, do not initialize `WP_Filesystem`. Instead, add documentation to the `generate_pdf` method and/or the new filter specifying that 3rd-party code wanting to use `WP_Filesystem` needs to make sure it is initialized. That way it can be done at the level of the `POST` request, and they can properly handle the [intricacies](https://codex.wordpress.org/Filesystem_API#Tips_and_Tricks) of the initialization.

This way, VIP sites could enable the filter and handle the initialization. I'm assuming that typical VIP have the creds stored in constants? In which case, using a solution such as the one you wrote should work fine, you would just have to add it to hte individual site(s).

Does this make sense? Any thoughts? I'm just very reluctant to add code to enable `WP_Filesystem` by default without making it work in all cases, and I don't see how we can properly initialize it in this plugin...